tapdb/ignore_tree.go

```diff
@@ -0,0 +1,380 @@
+			if err != nil {
+				return err
+			}
+
+			// To insert the universe leaf below, we'll need both
+			// the db primary key for the asset genesis, and also
+			// the raw bytes of the outpoint to be ignored.
+			assetGenID, err := db.LookupGenesisID(
+				ctx, ignoreTup.ID[:],
+			)
+			if err != nil {
+				return fmt.Errorf("error looking up genesis "+
+					"ID for asset %v: %w", ignoreTup.ID,
+					err)
+			}
+			ignorePointBytes, err := encodeOutpoint(
+				tup.IgnoreTuple.Val.OutPoint,
+			)
+			if err != nil {
+				return err
+			}
+
+			// With the leaf inserted into the tree, we'll now
+			// create the universe leaf that references the SMT
+			// leaf.
+			universeRootID, err := db.UpsertUniverseRoot(
+				ctx, NewUniverseRoot{
+					NamespaceRoot: namespace,
+					GroupKey:      groupKeyBytes,
+					ProofType: sqlStr(
+						id.ProofType.String(),
+					),
+				},
+			)
+			if err != nil {
+				return err
+			}
+
+			scriptKey := ignoreTup.ScriptKey
+			err = db.UpsertUniverseLeaf(ctx, UpsertUniverseLeaf{
+				AssetGenesisID:    assetGenID,
+				ScriptKeyBytes:    scriptKey.SchnorrSerialized(), //nolint:lll
+				UniverseRootID:    universeRootID,
+				LeafNodeKey:       smtKey[:],
+				LeafNodeNamespace: namespace,
+				MintingPoint:      ignorePointBytes,
+			})
+			if err != nil {
+				return err
+			}
+		}
+
+		// Fetch the final tree root after all insertions.
+		finalRoot, err := tree.Root(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Next, for each inserted tuple, generate its inclusion proof
+		// from the final tree and build its AuthenticatedIgnoreTuple.
+		for _, tup := range tuples {
+			smtKey := tup.IgnoreTuple.Val.Hash()
+			proof, err := tree.MerkleProof(ctx, smtKey)
+			if err != nil {
+				return err
+			}
+
+			authTup := universe.AuthenticatedIgnoreTuple{
+				SignedIgnoreTuple: tup,
+				InclusionProof:    proof,
+				IgnoreTreeRoot:    finalRoot,
+			}
+
+			finalResults = append(finalResults, authTup)
+		}
+
+		return nil
+	})
+	if txErr != nil {
+		return lfn.Err[[]universe.AuthenticatedIgnoreTuple](txErr)
+	}
+
+	return lfn.Ok(finalResults)
+}
+
+// Sum returns the sum of the ignore tuples for the given asset.
+func (it *IgnoreUniverseTree) Sum(ctx context.Context,
+	spec asset.Specifier) universe.SumQueryResp {
+
+	// Derive identifier from the asset.Specifier.
+	id, err := specifierToIdentifier(spec)
+	if err != nil {
+		return lfn.Err[lfn.Option[uint64]](err)
+	}
+	namespace := id.String()
+
+	var sumValue uint64
+	var foundSum bool
+
+	readTx := NewBaseUniverseReadTx()
+	txErr := it.db.ExecTx(ctx, &readTx, func(db BaseUniverseStore) error {
+		tree := mssmt.NewCompactedTree(
+			newTreeStoreWrapperTx(db, namespace),
+		)
+
+		// Get the root of the tree to retrieve the sum.
+		root, err := tree.Root(ctx)
+		if err != nil {
+			return err
+		}
+
+		// If root is empty, return empty sum.
+		//
+		// TODO(roasbeef): verify behavior
+		if root.NodeHash() == mssmt.EmptyTreeRootHash {
```

Comment:
Or, if we expect there to be no zero amount leaves in the tree, could also just do `root.NodeSum() == 0`.

Reply:
Can remove this TODO as it's verified with the unit tests, and also the impl: https://github.com/lightninglabs/taproot-assets/blob/581c8819aa95093494064f42f40adf700e3e4149/mssmt/store.go#L199-L201

https://github.com/lightninglabs/taproot-assets/blob/581c8819aa95093494064f42f40adf700e3e4149/tapdb/mssmt.go#L365-L369